// @threlte/core/dist/hooks/useFrame.js
import { getContext, onDestroy } from 'svelte';
import { readable, writable } from 'svelte/store';
import { browser } from '../lib/browser';
/**
 * Adds a handler to threltes unified render loop.
 *
 * `start` and `stop` functions are returned and the options allow
 * setting the handler to not start automatically.
 *
 * Provide an order to have your framehandlers running in a specific order.
 * Handlers are ordered from low to high when executing.
 * @param {(ctx: ThrelteContext, delta: number) => void} fn callback function
 * @param {ThrelteUseFrameOptions} options options
 * @returns {ThrelteUseFrame}
 */
export const useFrame = (fn, options) => {
    if (!browser) {
        return {
            start: () => undefined,
            stop: () => undefined,
            started: readable(false)
        };
    }
    const renderCtx = getContext('threlte-internal-context');
    if (renderCtx === undefined) {
        throw new Error('No Threlte context found, are you using this hook inside of <Canvas>?');
    }
    const invalidate = options?.invalidate ?? true;
    const handler = {
        fn,
        order: options?.order,
        debugFrameloopMessage: options?.debugFrameloopMessage,
        invalidate
    };
    const started = writable(false);
    const stop = () => {
        if (invalidate) {
            renderCtx.autoFrameHandlers.delete(handler);
        }
        else {
            renderCtx.manualFrameHandlers.delete(handler);
        }
        renderCtx.allFrameHandlers.delete(handler);
        started.set(false);
    };
    const start = () => {
        if (invalidate) {
            renderCtx.autoFrameHandlers.add(handler);
        }
        else {
            renderCtx.manualFrameHandlers.add(handler);
        }
        renderCtx.allFrameHandlers.add(handler);
        renderCtx.allFrameHandlersNeedSortCheck = true;
        started.set(true);
    };
    if (options?.autostart ?? true) {
        start();
    }
    onDestroy(() => {
        stop();
    });
    return {
        start,
        stop,
        started: {
            subscribe: started.subscribe
        }
    };
};
